File: QT-firewall/firewall-test/confignatrules.cpp
#include "confignatrules.h"
#include "iostream"

ConfigNATRules::ConfigNATRules(const std::string &devIP, const std::string &fw_mac, QObject *parent=nullptr)
                                :FwIP(devIP),fwMAC(fw_mac),QCommInfo(parent)
{
    fileOP = new FileOperation();
    this->setBindInfo(getLocal_IP(),30333);
}

ConfigNATRules::~ConfigNATRules()
{
    if(fileOP != nullptr)
        delete fileOP;
    fileOP = nullptr;
}

static std::string fileStr("");
static bool fileFlag = true;

void ConfigNATRules::ChangeSNAT(const std::string &ethName, const std::string &ethIP, const std::string &srcIP, bool addOrdel)
{
    std::string bridge_rule = "brctl delif br0 "+ethName;
    std::string ethIP_rule = "ifconfig "+ethName+" "+ethIP+" up";
    std::string rule = "iptables -t nat -A POSTROUTING -s "+srcIP+" -o br0 -j SNAT --to-source "+FwIP;

    std::string pro("NAT SNAT");
    std::string flag("NAT1");
    if(addOrdel)
    {
        fileFlag = true;
        std::string configInfo = flag+bridge_rule+" && "+ethIP_rule+" && "+rule;
        this->sendCmd(FwIP,22222,CommInfo::iptablesCmdTypeByte,configInfo);
    }
    else
    {
        flag.assign("NAT0");
        fileFlag = false;
        std::string configInfo = flag+bridge_rule+" && "+ethIP_rule+" && "+rule;
        this->sendCmd(FwIP,22222,CommInfo::iptablesCmdTypeByte,configInfo);
        std::string cmd = "brctl addif br0 "+ethName + " && " + "ifconfig "+ethName+" 0.0.0.0 up";
        this->sendCmd(FwIP,22222,CommInfo::commonShellCmdTypeByte,cmd);
    }

    fileStr = pro+" "+fwMAC+" "+ethName+" "+ethIP+" "+srcIP;
}

void ConfigNATRules::ChangeDNAT(const std::string &originDstIP, const std::string &originDstPort, const std::string &mapIP, const std::string &mapPort, bool addOrdel)
{
    std::string pro("NAT DNAT");
    std::string odp = " --dport "+originDstPort;
    std::string mdp = " --dport "+mapPort;
    if(originDstPort == "any")
        odp.assign("");
    if(mapPort == "any")
        mdp.assign("");
    std::string flag("NAT1");
    if(addOrdel)
    {
        fileFlag = true;
    }
    else
    {
        fileFlag = false;
        flag.assign("NAT0");
    }
    std::string config = flag+"iptables -t nat -A PREROUTING -d "+originDstIP+odp+" -j DNAT --to-destination "+mapIP+mdp;
    fileStr = pro+" "+fwMAC+" "+originDstIP+" "+originDstPort+" "+mapIP+" "+mapPort;
    std::cout<<config<<std::endl;

    this->sendCmd(FwIP,22222,CommInfo::iptablesCmdTypeByte,config);
}

//void ConfigNATRules::ClearAllNATRules()
//{
//    std::string cmd = "iptables -t nat -F";
//    this->sendCmd(FwIP,22222,CommInfo::commonShellCmdTypeByte,cmd);
//}

void ConfigNATRules::handResultData(const bool result, const char *commandTypeByte, const char *commandBytes, const int commandLength, const QString &srcIp, const int srcPort)
{
    if (result)
    {
        if (*commandTypeByte == *CommInfo::iptablesCmdTypeByte)
        {
            std::string data(commandBytes,commandLength);
            //std::cout << data << std::endl;
            if(data == "success")
            {
                if(fileFlag)
                    fileOP->addRule(fileStr);
                else
                    fileOP->deleteRule(fileStr);
                emit configResult(true);
            }

            if(data == "fail")
                emit configResult(false);
        }
    }
}

File: 防火墙项目/firewallclient-nodejs/classes/configNATRules.js
var sendInfo = require('../tools/sendInfo.js');
var dbOperations = require('../tools/dbOperations.js');
var configInfo = require('../tools/listenConfigInfo.js');

/**
 * 配置源地址映射规则
 * Author：于仁飞
 * @param {*防火墙对象，无IP模式下不能配置} dev 
 * @param {*网口} ethName 
 * @param {*网口对应设备的IP地址} devIp 
 * @param {*网口IP地址} ethIp 
 * @param {*网口IP对应的掩码}netmask
 * @param {*true为添加,false为删除} addOrDelete 
 */

function configSNATRules(dev, ethName, devIp, ethIp, netmask, addOrDelete) {
    if (dev.getDevIP() == "0.0.0.0")
        return false;
    var configEthBr, configEthIp, sqlRule, SNATRules;
    var rule = "iptables -t nat -A POSTROUTING -s " + devIp + " -o br0 -j SNAT --to-source " + dev.getDevIP();
    if (addOrDelete) {
        configEthBr = "brctl delif br0 " + ethName;
        configEthIp = "ifconfig " + ethName + " " + ethIp + " netmask " + netmask + " up";
        SNATRules = "NAT1" + configEthBr + " && " + configEthIp + " && " + rule;
        sqlRule = "INSERT INTO snat VALUES('" + dev.getDevMAC() + "','" + devIp + "','" + ethName + "','" + ethIp + "','" + netmask + "','" + dev.getDevIP() + "')";
    }
    if (!addOrDelete) {
        configEthBr = "brctl addif br0 " + ethName;
        configEthIp = "ifconfig " + ethName + " 0.0.0.0 up";
        SNATRules = "NAT0" + configEthIp + " && " + configEthBr + " && " + rule;
        sqlRule = "DELETE FROM snat where(fwMAC='" + dev.getDevMAC() + "' and origin_devIP='" + devIp + "' and EthName='" + ethName + "' and EthIP='" + ethIp + "' and netmask='" + netmask + "' and NATIP='" + dev.getDevIP() + "')";
    }
    console.log(SNATRules);
    console.log(sqlRule);
    dbOperations.dbOperation(sqlRule);
    sendInfo.sendConfigInfo(dev.getDevIP(), SNATRules);
    return configInfo.listenConfigInfo();
}
exports.configSNATRules = configSNATRules;

/**
 * 配置目的地址映射规则
 * Author：于仁飞
 * @param {*防火墙设备对象，无IP模式下不能配置} dev 
 * @param {*原目的IP地址} originIp 
 * @param {*原目的端口，any表示任意端口} originPort 
 * @param {*映射IP地址} mapIp 
 * @param {*映射端口，any表示任意端口} mapPort 
 * @param {*true为添加,false为删除} addOrDelete 
 */
function configDNATRules(dev, originIp, originPort, mapIp, mapPort, addOrDelete) {
    if (dev.getDevIP() == "0.0.0.0")
        return false;
    var preRule, sqlRule, DNATRules;
    if (originPort != "any" & mapPort != "any") {
        preRule = "iptables -t nat -A PREROUTING -d" + " " + originIp + " " + "-p tcp --dport " + originPort +
            " -j DNAT --to-destination " + mapIp + ":" + mapPort;
    }
    if (originPort != "any" & mapPort == "any") {
        preRule = "iptables -t nat -A PREROUTING -d" + " " + originIp + " " + "-p tcp --dport " + originPort +
            " -j DNAT --to-destination " + mapIp;
    }
    if (originPort == "any" & mapPort != "any") {
        preRule = "iptables -t nat -A PREROUTING -d" + " " + originIp + " " + "-p tcp " + " -j DNAT --to-destination " + mapIp + ":" + mapPort;
    }
    if (originPort == "any" & mapPort == "any") {
        preRule = "iptables -t nat -A PREROUTING -d" + " " + originIp + " " + "-p tcp " + " -j DNAT --to-destination " + mapIp;
    }
    if (addOrDelete) {
        DNATRules = "NAT1" + preRule;
        sqlRule = "INSERT INTO dnat values('" + dev.getDevMAC() + "','" + originIp + "','" + originPort + "','" + mapIp + "','" + mapPort + "')";
    }
    if (!addOrDelete) {
        DNATRules = "NAT0" + preRule;
        sqlRule = "DELETE FROM dnat where(fwMAC='" + dev.getDevMAC() + "' and origin_dstIP='" + originIp + "' and origin_dport='" + originPort + "' and map_IP='" + mapIp + "' and map_port='" + mapPort + "')";
    }
    dbOperations.dbOperation(sqlRule);
    sendInfo.sendConfigInfo(dev.getDevIP(), DNATRules);
    return configInfo.listenConfigInfo();
}
exports.configDNATRules = configDNATRules;